### PR TITLE
Tweak predictProblems logic, add test for it

### LINKED ISSUE #2114: Import: Missing name not caught when ID is also missing
## Description
There is error-checking in the importer that makes sure we don't try to import (or even validate on the server) rows that are invalid in some way. One such way is when we try to import a person which has no first name or last name. The same checks also warn against importing without an ID. But it seems that when you configure an import without an ID, the warning about missing ID obscures the deal-breaker error that should be displayed because there is no name.

## Steps to reproduce
1. Go to https://app.dev.zetkin.org/organize/1/people
2. Click on "Create"
3. Click on "Import people"
4. Pick a file that contains names but not IDs (see example CSV below)
5. Configure first name and last name column
6. Click preview

### Example CSV
```csv
FNAME,LNAME,EMAIL
Clara,Zetkin,clara@zetkin.org
Clara,,clara@example.com
```

### Expected Behaviour
A red error should be displayed (instead of the yellow warning) letting the user know that there is no ID or name on one row. Internally in the code, this problem is called `ImportProblemKind.MISSING_ID_AND_NAME`.

### Actual Behaviour
A yellow warning about missing ID is displayed, but can be bypassed by the user (even though the import will fail on the next step).

## Screenshots (if you have any)
Note how the HTTP request has failed, because the client tried to make the request even though it should have found that there was a missing name and stopped.

image

## Proposed solution
Add a test case for this in `predictProblems.spec.ts` and tweak `predictProblems()` to catch the problem.

### PR DESCRIPTION
**Pull request by kaulfield23**
## Description
This PR fixes `predictProblems` logic where it shows an error when name value is missing.

## Screenshots

https://github.com/user-attachments/assets/be2d365d-5709-4ad4-826f-b43600ca9e3b

## Changes

* Adds test to `predictProblems.spec.ts`
* Changes `predictProblems` logic

## Related issues
Resolves #2114

### CODE REVIEW

File: src/features/import/utils/problems/predictProblems.ts [hunk 1]
```diff
@@ -75,6 +63,7 @@ export function predictProblems(
   customFields.forEach((field) => {
     customFieldsBySlug[field.slug] = field;
   });
+  let isMissingValue = false;
```
Comment (ziggabyte) [inline src/features/import/utils/problems/predictProblems.ts:66]: I am not really sure this variable is doing so much 😅 If I remove it from the code, the only difference is that I get both unconfigured ID error and "not all rows have identifiers" error, instead of just the latter.

File: src/features/import/utils/problems/predictProblems.ts [hunk 2]
```diff
@@ -133,11 +122,13 @@ export function predictProblems(
             }
           }
           hadImpact = true;
+        } else {
+          isMissingValue = true;
```
Comment (ziggabyte) [inline src/features/import/utils/problems/predictProblems.ts:126]: This made me a bit confused - this will be set to true if any "cell" in the sheet is empty, no matter if it is name/id or not.

**Review by ziggabyte — CHANGES_REQUESTED** [review]
Hi Haeju!! :D 

I have played around with this a bit and read the code, and even though it does work for the case you are testing for - this solution, I think, can be made better because it has an unexpected buggy consequence... 

If I run your code and import a file without giving selecting a first name, last name or id column, i get "unexpected error". If i do the same thing on `main`, I get the error "you have not configured identifying columns", which is what should happen... 

So, maybe experiment some more and see that it has no unexpected consequences!

File: src/features/import/utils/problems/predictProblems.ts (hunk 1 above)
Comment (kaulfield23) [reply src/features/import/utils/problems/predictProblems.ts:66]: This is just for removing unconfigured ID error(yellow box). Thought that it is weird to show yellow box with text "I understand and want to proceed anyway". Should I remove it?

Comment (richardolsson) [reply src/features/import/utils/problems/predictProblems.ts]: I think it makes sense to do what you've tried to do @kaulfield23, but I think the best way to do it is to write a test that reproduces the problem you found, and then fix it.

**Comment by richardolsson** [comment]
Nice work @kaulfield23. Can you also write tests for the edge cases you've handled?

File: src/features/import/utils/problems/predictProblems.ts (hunk 1 above)
Comment (kaulfield23) [reply src/features/import/utils/problems/predictProblems.ts]: @richardolsson Okay! will do 👍

Comment (kaulfield23) [reply src/features/import/utils/problems/predictProblems.ts]: I think that case is already covered with the test I wrote ( test : returns problem when either the first name or last name value is missing ) . And the case that ziggi found already covered with another test that already exists ( test  : returns problem when ID or name are not configured )

Is there a case that I have not added? @richardolsson

**Review by richardolsson — CHANGES_REQUESTED** [review]
I've found a regression in this PR (i.e. a bug that was introduced in this PR that doesn't exist on `main`). It may be the same thing that @ziggabyte pointed out in his previous review, and it still remains after the most recent commit. Below are instructions for how to reproduce it (and separate a comment on how I think it should be tested).

I think this needs to be fixed before this can be merged. But then I think we also need better testing in the future, and I've created a separate issue for this: #2156.

## Steps to reproduce
Use a CSV file with this content (note the missing first name on the last row):

```csv
ID,FIRST,LAST,EMAIL
4,Clara,Zetkin,clara1@example.com
,Clara,Zetin,clara2@example.com
5,,Zetkin,clara3@exaemple.com
```

Configure only the first name and the email address, not the ID or last name columns, and then click "VALIDATE"

image

### Expected results
The user should be presented with an error message indicating that not enough identifying columns have been configured. This should be caught by `predictProblems()` and prevent any API requests from being made. This currently works on `main`.

image

### Actual results
The user is presented with an "unknown" error message. An API request is made to the /preview API which immediately fails. This likely happens because `predictProblems()` finds no errors.

image

File: src/features/import/utils/problems/predictProblems.spec.ts [hunk 3]
```diff
@@ -459,6 +485,65 @@ describe('predictProblem()', () => {
...
+      ],
+    });
+
+    const problems = predictProblems(sheet, 'SE', []);
+    expect(problems).toEqual([
+      {
+        kind: ImportProblemKind.UNCONFIGURED_ID_AND_NAME,
+      },
+      {
+        kind: 'NO_IMPACT',
```
Comment (kaulfield23) [inline src/features/import/utils/problems/predictProblems.spec.ts:542]: I added a new test where column is not configured. And I got this 'NO_IMPACT' together in problem result. Is this fine?

File: src/features/import/utils/problems/predictProblems.ts [hunk 4]
```diff
@@ -134,26 +118,57 @@ export function predictProblems(
...
-        rowProblemByKind[ImportProblemKind.MISSING_ID_AND_NAME].indices.push(
-          rowIndex
-        );
+    // When id is missing and first and last name columns configured but it is missing name value or
+    // when id is missing and one of name column configured
+    const problemWithIdColumn =
+      sheetHasId &&
+      !rowHasId &&
+      (sheetHasFirstName || sheetHasLastName) &&
+      (missingFirstOrLastName || !sheetHasFirstName || !sheetHasLastName);
```
Comment (kaulfield23) [inline src/features/import/utils/problems/predictProblems.ts:139]: Handle a case where ID is missing when only one of name column is configured.
For example
```
 columns: [
        {
          idField: 'id',
          kind: ColumnKind.ID_FIELD,
          selected: true,
        },
        {
          field: 'first_name',
          kind: ColumnKind.FIELD,
          selected: true,
        },
      ],
      rows: [{ data: [null, 'Clara'] }],

Comment (kaulfield23) [inline src/features/import/utils/problems/predictProblems.ts:139]: Handle a case where ID is missing when only one of name column is configured.
For example
```
 columns: [
        {
          idField: 'id',
          kind: ColumnKind.ID_FIELD,
          selected: true,
        },
        {
          field: 'first_name',
          kind: ColumnKind.FIELD,
          selected: true,
        },
      ],
      rows: [{ data: [null, 'Clara'] }],
```

File: src/features/import/utils/problems/predictProblems.spec.ts [hunk 5]
```diff
@@ -459,6 +485,65 @@ describe('predictProblem()', () => {
...
+    ]);
+  });
+
+  it('returns correct problem when column is not configured', () => {
+    const sheet = makeFullSheet({
+      columns: [
+        {
+          kind: ColumnKind.UNKNOWN,
+          selected: false,
+        },
```
Comment (richardolsson) [inline src/features/import/utils/problems/predictProblems.spec.ts:528]: The columns here does not match the data. The data contains three columns, but there is only one column here. There should be three columns here.

If this test is intended to accommodate [my comment from my previous review](https://github.com/zetkin/app.zetkin.org/pull/2117/files#r1776347820), there seems to have been a misunderstanding.

This test should contain three columns, the same three columns that the previous test, but with the first one (the one where first name is) de-selected.

```suggestion
        {
          kind: ColumnKind.UNKNOWN,
          selected: false,
        },
        {
          field: 'last_name',
          kind: ColumnKind.FIELD,
          selected: true,
        },
        {
          field: 'email',
          kind: ColumnKind.FIELD,
          selected: true,
        },
```

File: src/features/import/utils/problems/predictProblems.spec.ts (hunk 3 above)
Comment (richardolsson) [reply src/features/import/utils/problems/predictProblems.spec.ts:542]: This should not be returning `NO_IMPACT`. I believe that happens because the test you have written contains _no selected columns_.

**Review by richardolsson — CHANGES_REQUESTED** [review]
I believe this is working mostly as it should, but there is a problem with one of the new tests IMO.

The logic in the application code is also getting really complex and hard to read, but if we can make sure it works first, and all the tests are correct, we can refactor and simplify the logic later and rely on the tests.

**Comment by kaulfield23** [comment]
It says deployment has failed but tests are passed so I guess it is fine :)

**Comment by richardolsson** [comment]
Your stamina on this has been impressive @kaulfield23! I believe you may have finally fixed the bug?

I'm so sorry to say this, but I don't think we should merge this now. There is a separate PR, #2472, which replaces this logic entirely in order to utilize new capabilities in the backend. Merging this PR will cause conflicts when time comes to merge #2472, and the way it will be resolved is by removing all changes from this PR and replacing them with the ones in #2472, so there will be no benefit from having merged this.

I will close this PR, and I hope that you are willing to find some other issue to work on. Hope to see you at a hackathon or code camp soon!
